wvdial.1
.TH WVDIAL 1 "December 2005" "WvDial"
.\"
.SH NAME
wvdial \- PPP dialer with built-in intelligence.
.\"
.SH SYNOPSIS
.B wvdial
[
.I OPTIONS
]
[
.I SECTION
] ...
.\"
.SH DESCRIPTION
.B wvdial
is an intelligent PPP dialer, which means that it dials a modem and
starts PPP in order to connect to the Internet.  It is something like the
.BR chat (8)
program, except that it uses heuristics to guess how to dial and log
into your server rather than forcing you to write a login script.
.PP
When
.B wvdial
starts, it first loads its configuration from
.IR /etc/wvdial.conf
and
.IR ~/.wvdialrc
which contains basic information about the modem port, speed, and init
string, along with information about your Internet Service Provider (ISP),
such as the phone number, your username, and your password.
.PP
Then it initializes your modem and dials the server and waits for a 
connection (a CONNECT string from the modem).  It understands and responds 
to typical connection problems (like BUSY and NO DIALTONE).
.PP
Any time after connecting, 
.B wvdial
will start PPP if it sees a PPP sequence from the server.  Otherwise,
it tries to convince the server to start PPP by doing the following:
.IP \(bu 4
responding to any login/password prompts it sees;
.IP \(bu
interpreting \(lqchoose one of the following\(rq-style menus;
.IP \(bu
eventually, sending the word \(lqppp\(rq (a common terminal server command).
.PP
If all of this fails,
.B wvdial
just runs
.BR pppd (8)
and hopes for the best. It will bring up the connection, and then wait
patiently for you to drop the link by pressing CTRL\-C.
.\"
.SH OPTIONS
Several options are recognized by
.BR wvdial .
.TP
.B \-c, \-\-chat
Run wvdial as a chat replacement from within pppd, instead of the more
normal method of having wvdial negotiate the connection and then call
pppd.
.TP
.B \-C, \-\-config=CONFIGFILE
Run wvdial with CONFIGFILE as the configuration file, instead of
.IR /etc/wvdial.conf .
This is mainly useful only if you want to have per-user
configurations, or you want to avoid having dial-up information (usernames,
passwords, calling card numbers, etc.) in a system wide configuration file.
.TP
.B \-n, \-\-no\-syslog
Don't output debug information to the syslog daemon (only useful together
with \-\-chat).
.PP
.B wvdial
is normally run without command line options, in which case it reads
its configuration from the
.I "[Dialer Defaults]"
section of
.IR /etc/wvdial.conf .
(The configuration file is described in more detail in
.BR wvdial.conf (5)
manual page.)
.PP
One or more 
.IR SECTION s
of
.I
/etc/wvdial.conf
may be specified on the command line.  Settings in these sections will
override settings in
.IR "[Dialer Defaults]" .
.PP
For example, the command:
.RS
.B wvdial phone2
.RE
will read default options from the
.I "[Dialer Defaults]"
section, then override any or all of the options with those found in the
.I "[Dialer phone2]"
section.
.PP
If more than one section is specified, they are processed in the order they
are given.  Each section will override all the sections that came before it.
.PP
For example, the command:
.RS
.B wvdial phone2 pulse shh
.RE
will read default options from the
.I "[Dialer Defaults]"
section, then override any or all of the options with those found in the
.I "[Dialer phone2]"
section, followed by the
.I "[Dialer pulse]"
section, and lastly the
.I "[Dialer shh]"
section.
.PP
Using this method, it is possible to easily configure
.B wvdial
to switch between different internet providers, modem init strings,
account names, and so on without specifying the same configuration
information over and over.
.\"
.SH BUGS
\(lqIntelligent\(rq programs are frustrating when they don't work right.
This version of
.B wvdial
has only minimal support for disabling or overriding its \(lqintelligence\(rq,
with the \(lqStupid Mode\(rq, \(lqLogin Prompt\(rq, and \(lqPassword 
Prompt\(rq options.
So, in general if you have a nice ISP, it will probably work,
and if you have a weird ISP, it might not.
.PP
Still, it's not much good if it doesn't work for you, right?  Don't be fooled
by the fact that
.B wvdial
finally made it to version 1.00; it could well contain many bugs and 
misfeatures.  Let us know if you have problems by sending e-mail to
.BR <wvdial\-list@lists.nit.ca> .
.PP
You may encounter some error messages if you don't have write access to
.B /etc/ppp/pap\-secrets
and
.BR /etc/ppp/chap\-secrets .
Unfortunately, there's really no nice way around this yet.
.SH FILES
.TP
/etc/wvdial.conf
Configuration file which contains modem, dialing, and login
information. See 
.BR wvdial.conf (5).
.TP
/dev/ttyS*
Serial port devices.
.TP
/etc/ppp/peers/wvdial
Required for correct authentication in pppd version 2.3.0 or newer.
.TP
/etc/ppp/{pap,chap}\-secrets
Contains a list of usernames and passwords used by pppd for authentication.
.B wvdial
maintains this list automatically.
.\"
.SH AUTHORS
Dave Coombs and Avery Pennarun for Net Integration Technologies.
We would also like to thank SuSE and
RedHat for adding a number of various cool features to
.BR wvdial .
Thanks guys!
.\"
.SH SEE ALSO
.BR wvdial.conf (5),
.BR wvdialconf (1),
.BR pppd (8),
.BR chat (8).



wvdial.cc
/*
 * Worldvisions Weaver Software:
 *   Copyright (C) 1997-2005 Net Integration Technologies, Inc.
 *
 * Standalone WvDial program, for testing the WvDialer class.
 *
 * Created:	Sept 30 1997		D. Coombs
 */

#include "wvargs.h"
#include "wvdialer.h"
#include "wvver.h"
#include "wvlog.h"
#include "wvlogrcv.h"
#include "wvlogfile.h"
#include "wvsyslog.h"
#include "wvcrash.h"

#include <signal.h>
#include <sys/wait.h>
#include <unistd.h>

volatile bool want_to_die = false;


// use no prefix string for app "Modem", and an arrow for everything else.
// This makes the output of the wvdial application look nicer.
class WvDialLogger : public WvLogConsole
/**************************************/
{
public:
    WvDialLogger() : WvLogConsole(dup(2)) // log to stderr (fd 2)
        { }

protected:
    virtual void _make_prefix();
};


void WvDialLogger::_make_prefix()
/*******************************/
{
    WvString name = last_source;
    if(name == "WvDial Modem") 
    {
	prefix = "";
	prelen = 0;
    } 
    else 
    {
	prefix = "--> ";
	prelen = 4;
    }
}

static void signalhandler(int sig)
/***********************************/
{
    fprintf(stderr, "Caught signal %d:  Attempting to exit gracefully...\n", sig);
    want_to_die = true;
    signal(sig, SIG_DFL);
}


bool haveconfig = false;
static bool config_cb(WvStringParm value, void *userdata)
{
    WvConf *cfg = reinterpret_cast<WvConf*>(userdata);
    if (!access(value, F_OK))
    {
	cfg->load_file(value);
	haveconfig = true;
	return true;
    }
    fprintf(stderr, "Cannot read `%s'\n", value.cstr());
    return false;
}


int main(int argc, char **argv)
/********************************/
{
#if DEBUG
    free( malloc( 1 ) ); // for electric fence
#endif
    
    WvDialLogger 	rc;
    WvSyslog		*syslog = NULL;
    WvLogFile           *filelog = NULL;
    UniConfRoot         uniconf("temp:");
    WvConf              cfg(uniconf);
    WvStringList	sections;
    WvStringList	cmdlineopts;
    WvLog		log( "WvDial", WvLog::Debug );
    WvString		homedir = getenv("HOME");
    
    bool chat_mode = false;
    bool write_syslog = true;
    
    signal(SIGTERM, signalhandler);
    signal(SIGINT, signalhandler);
    signal(SIGHUP, signalhandler);

    WvArgs args;
    args.set_version("WvDial " WVDIAL_VER_STRING "\n"
		     "Copyright (c) 1997-2005 Net Integration Technologies, "
		     "Inc.");
    args.set_help_header("An intelligent PPP dialer.");
    args.set_help_footer("Optional SECTION arguments refer to sections in "
			 "configuration file (usually)\n"
			 "/etc/wvdial.conf, $HOME/.wvdialrc or the file "
			 "specified by --config.\n"
			 "Specified sections are all read, with later ones "
			 "overriding previous ones.\n"
			 "Any options not in the listed sections are taken "
			 "from [Dialer Defaults].\n"
			 "\n"
			 "Also, optional OPTION=value parameters allow you "
			 "to override options within\n"
			 "the configuration files.\n");
    args.set_email("<wvdial-list@lists.nit.ca>");

    args.add_option('C', "config",
		    "use configfile instead of /etc/wvdial.conf",
		    "configfile", WvArgs::ArgCallback(&config_cb), &cfg);
    args.add_set_bool_option('c', "chat",
			     "used when running wvdial from pppd", chat_mode);
    args.add_reset_bool_option('n', "no-syslog",
			       "don't send output to SYSLOG", chat_mode);
    args.add_optional_arg("SECTION", true);
    args.add_optional_arg("OPTION=value", true);

    WvStringList remaining_args;
    args.process(argc, argv, &remaining_args);

    {
	WvStringList::Iter i(remaining_args);
	for (i.rewind(); i.next(); )
	{
	    if (strchr(i(), '=' ))
		cmdlineopts.append(new WvString(i()),true);
	    else
		sections.append(new WvString("Dialer %s", i()), true);
	}
    }

    if (sections.isempty())
	sections.append(new WvString("Dialer Defaults"), true);

    if( !haveconfig)
    {
	// Load the system file first...
	WvString stdconfig("/etc/wvdial.conf");
		
	if (!access(stdconfig, F_OK))
	    cfg.load_file(stdconfig);
	
	// Then the user specific one...
	if (homedir)
        {
	    WvString rcfile("%s/.wvdialrc", homedir);
			
	    if (!access(rcfile, F_OK))
		cfg.load_file(rcfile);
	}
    }
    
    // Inject all of the command line options on into the cfg file in a new
    // section called Command-Line if there are command line options.
    if (!cmdlineopts.isempty()) 
    {
        WvStringList::Iter i(cmdlineopts);
        for (i.rewind();i.next();)
        {
            char *name = i().edit();
            char *value = strchr(name,'=');
	    
            // Value should never be null since it can't get into the list
            // if it doesn't have an = in i()
            // 
            *value = 0;
	    value++;
            name = trim_string(name);
            value = trim_string(value);
            cfg.set("Command-Line", name, value);
        }
        sections.prepend(new WvString("Command-Line"), true);
    }
    
    if(!cfg.isok()) 
    {
	return 1;
    }
    
    if (chat_mode) 
    { 
	if (write_syslog) 
	{ 
	    WvString buf("wvdial[%s]", getpid()); 
	    syslog = new WvSyslog( buf, false, WvLog::Debug2, 
				   WvLog::Debug2 ); 
	} 
	else 
	{ 
	    // Direct logging to /dev/null as otherwise WvLog hasn't any 
	    // receivers and thus will use WvLogConsole to log to stderr. 
	    // That can disturb the communication with the modem on 
	    // stdin/stdout. - Fixes a bug reported by SUSE on 04/05/04
	    filelog = new WvLogFile( "/dev/null", WvLog::Debug2 ); 
	} 
    }
    
    WvDialer dialer(cfg, &sections, chat_mode);
    
    if (!chat_mode)
	if (dialer.isok() && dialer.options.ask_password)
	    dialer.ask_password();
    
    if (dialer.dial() == false)
	return  1;
    
    while (!want_to_die && dialer.isok() 
	   && dialer.status() != WvDialer::Idle) 
    {
	dialer.select(100);
	dialer.callback();
    }
    
    int retval;
    
    if (want_to_die)
    {
	// Probably dieing from a user signal
        retval = 2;
    }
    
    if ((dialer.status() != WvDialer::Idle) || !dialer.isok()) 
    {
	retval = 1;
    } 
    else 
    {
	retval = 0;
    }
    
    dialer.hangup();
    
    WVRELEASE(filelog);
    delete syslog;

    return(retval);
}
